src/es6-module-utils.ts

```diff
@@ -0,0 +1,444 @@
+        } else {
+          trialName = `${trialName}$1`;
+        }
+      }
+    }
+    moduleExports.set(name, exportName);
+  }
+  return exportName;
+}
+
+export function hasDefaultModuleExport(node: babel.Node): boolean {
+  let hasDefaultModuleExport = false;
+  traverse(node, {
+    noScope: true,
+    ExportDefaultDeclaration: {
+      enter(path: NodePath) {
+        hasDefaultModuleExport = true;
+        path.stop();
+      }
+    }
+  });
+  return hasDefaultModuleExport;
+}
+
+export function getModuleExportNames(node: babel.Node): Set<string> {
+  const exportedNames: string[] = [];
+  traverse(node, {
+    noScope: true,
+    ExportNamedDeclaration: {
+      enter(path: NodePath) {
+        const exportNode: babel.ExportNamedDeclaration = path.node as any;
+        exportedNames.push(
+            ...getModuleExportIdentifiers(...exportNode.specifiers));
+        exportedNames.push(
+            ...getModuleExportIdentifiers(exportNode.declaration));
+      }
+    }
+  });
+  return new Set(exportedNames);
+}
+
+function getModuleExportIdentifiers(...nodes: babel.Node[]): string[] {
+  const identifiers = [];
+  for (const node of nodes) {
+    if (babel.isArrayPattern(node)) {
+      identifiers.push(...getModuleExportIdentifiers(...node.elements));
+    }
+    if (babel.isClassDeclaration(node) || babel.isFunctionDeclaration(node) ||
+        babel.isVariableDeclarator(node)) {
+      identifiers.push(...getModuleExportIdentifiers(node.id));
+    }
+    if (babel.isExportSpecifier(node)) {
+      identifiers.push(...getModuleExportIdentifiers(node.exported));
+    }
+    if (babel.isIdentifier(node)) {
+      identifiers.push(node.name);
+    }
+    if (babel.isObjectPattern(node)) {
+      identifiers.push(...getModuleExportIdentifiers(...node.properties));
+    }
+    if (babel.isObjectProperty(node)) {
+      identifiers.push(...getModuleExportIdentifiers(node.value));
+    }
+    if (babel.isVariableDeclaration(node)) {
+      identifiers.push(...getModuleExportIdentifiers(...node.declarations));
+    }
+  }
+  return identifiers;
+}
+
+export async function reserveBundleModuleExportNames(
+    analyzer: Analyzer, manifest: BundleManifest) {
+  const es6ModuleBundles =
+      [...manifest.bundles]
+          .map(([url, bundle]) => ({url, bundle}))
+          .filter(({bundle}) => bundle.type === 'es6-module');
+  const analysis = await analyzer.analyze(es6ModuleBundles.map(({url}) => url));
+  for (const {url, bundle} of es6ModuleBundles) {
+    if (bundle.files.has(url)) {
+      const document = getAnalysisDocument(analysis, url);
+      for (const exportName of getModuleExportNames(
+               document.parsedDocument.ast as any)) {
+        getBundleModuleExportName({url, bundle}, url, exportName);
+      }
+    }
+  }
+}
+
+/**
+ * Utility class to rollup/merge ES6 modules code using rollup and rewrite
+ * import statements to point to appropriate bundles.
+ */
+export class Es6Rewriter {
```

Comment:
maybe move this to its own file? doesn't seem like a "utility" thing so much as one of the most central functional blocks of the whole PR :)

Reply:
Yeah I guess it grew a bit.  I'll move it after addressing the feedback.